File: backend/app/schemas/game.py
import uuid
from typing import Optional

from pydantic import BaseModel


class CreateGameRequest(BaseModel):
    nickname: str
    max_players: int = 6


class JoinGameRequest(BaseModel):
    nickname: str


class CreateGameResponse(BaseModel):
    room_code: str
    game_id: uuid.UUID
    player_id: uuid.UUID


class PlayerInfo(BaseModel):
    id: uuid.UUID
    nickname: str
    seat_order: int
    is_host: bool
    total_score: int
    is_connected: bool


class GameStateResponse(BaseModel):
    game_id: uuid.UUID
    room_code: str
    status: str
    num_players: int
    max_players: int
    current_round: int
    total_rounds: int
    current_phase: str
    dealer_seat: int
    current_player_seat: int
    trump_suit: Optional[str]
    trump_card: Optional[str]
    players: list[PlayerInfo]

File: frontend/src/types/game.ts
export type GameStatus = 'waiting' | 'in_progress' | 'finished'
export type GamePhase = 'waiting' | 'dealing' | 'bidding' | 'playing' | 'scoring'
export type Suit = 'C' | 'D' | 'H' | 'S' | 'none' | null

export interface PlayerInfo {
  id: string
  nickname: string
  seat_order: number
  is_host: boolean
  total_score: number
  is_connected: boolean
}

export interface GameState {
  game_id: string
  room_code: string
  status: GameStatus
  num_players: number
  max_players: number
  current_round: number
  total_rounds: number
  current_phase: GamePhase
  dealer_seat: number
  current_player_seat: number
  trump_suit: Suit
  trump_card: string | null
  players: PlayerInfo[]
}

export interface RoundState {
  hand: string[]
  bids: Record<string, number | null>
  tricks_won: Record<string, number>
  current_trick: Array<{ player_id: string; card: string }>
  last_trick_winner: string | null
  trick_winner_id: string | null  // set during the 1.5 s highlight window
}

// WebSocket message shapes
export interface WsMessage {
  type: string
  payload: unknown
}

export interface WsGameState extends WsMessage {
  type: 'game_state'
  payload: GameState
}

export interface WsRoundStarted extends WsMessage {
  type: 'round_started'
  payload: {
    round_number: number
    hand: string[]
    trump_card: string | null
    trump_suit: Suit
    dealer_seat: number
    first_bidder_seat: number
  }
}

export interface WsCardPlayed extends WsMessage {
  type: 'card_played'
  payload: { player_id: string; card: string }
}

export interface WsTrickComplete extends WsMessage {
  type: 'trick_complete'
  payload: { winner_id: string; cards: Array<{ player_id: string; card: string }> }
}

export interface WsRoundComplete extends WsMessage {
  type: 'round_complete'
  payload: { scores: Record<string, number>; cumulative_scores: Record<string, number> }
}

export interface WsGameComplete extends WsMessage {
  type: 'game_complete'
  payload: { final_scores: Record<string, number> }
}
